File: core/parsers/parser_registry.py
from __future__ import annotations

from pathlib import Path
from typing import Dict, Optional

from .ast_engine import parse_ast
from .call_graph_engine import build_call_graph
from .dependency_resolution_engine import resolve_dependencies
from .import_resolution_engine import resolve_imports
from .parser_budget_engine import ParserBudget, enforce_budget
from .parser_recovery_engine import recover_syntax
from .runtime_resolution_engine import resolve_runtime
from .api_resolution_engine import resolve_api_surface
from .framework_resolution_engine import resolve_frameworks
from .semantic_graph_engine import build_semantic_graph
from .symbol_resolution_engine import resolve_symbols
from .parser_output_engine import normalize_parser_output
from .formal_parser_grounding_engine import require_parser_evidence

_EXTENSION_LANGUAGE = {
    ".py": "python",
    ".js": "javascript",
    ".jsx": "javascript",
    ".ts": "typescript",
    ".tsx": "typescript",
    ".java": "java",
    ".kt": "kotlin",
    ".go": "go",
    ".rs": "rust",
    ".dart": "dart",
}


class ParserRegistry:
    @staticmethod
    def detect_language(path: str = "", hint: str = "") -> str:
        if hint:
            return hint.lower()
        ext = Path(path).suffix.lower()
        return _EXTENSION_LANGUAGE.get(ext, "text")

    @staticmethod
    def parse(
        source: str,
        path: str = "",
        language_hint: str = "",
        budget: Optional[ParserBudget] = None,
    ) -> Dict[str, object]:
        language = ParserRegistry.detect_language(path=path, hint=language_hint)
        bounded = enforce_budget(source, budget)
        recovered = recover_syntax(bounded, language)
        ast_tree = parse_ast(recovered, language)
        symbols = resolve_symbols(recovered, language, ast_tree=ast_tree)
        calls = build_call_graph(recovered, language)
        imports = resolve_imports(symbols, source_id=path or language)
        deps = resolve_dependencies(recovered, path=path)
        runtime = resolve_runtime(
            deps.get("dependencies", []),
            symbols.get("imports", []),
        )
        frameworks = resolve_frameworks(
            deps.get("dependencies", []),
            symbols.get("imports", []),
            symbols.get("decorators"),
        )
        api_surface = resolve_api_surface(recovered, language, path=path)
        parsed = {
            "language": language,
            "source_id": path or language,
            "ast": ast_tree,
            "symbols": symbols,
            "calls": calls,
            "imports": imports,
            "dependencies": deps,
            "runtime": runtime,
            "frameworks": frameworks,
            "api_surface": api_surface,
        }
        parsed["semantic_graph"] = build_semantic_graph(parsed)
        parsed["evidence"] = {
            "ast": bool(ast_tree.get("nodes")),
            "symbols": bool(symbols.get("symbols")),
            "calls": bool(calls.get("calls")),
            "dependencies": bool(deps.get("dependencies")),
            "tree_sitter": language != "python" and not ast_tree.get("parse_error"),
        }
        if ast_tree.get("parse_error"):
            parsed["evidence"]["parse_error"] = True
        normalized = normalize_parser_output(parsed)
        parsed.update(normalized)
        parsed["parser_grounding"] = require_parser_evidence(parsed)
        return parsed


def parse_source(
    source: str,
    path: str = "",
    language_hint: str = "",
    budget: Optional[ParserBudget] = None,
) -> Dict[str, object]:
    return ParserRegistry.parse(source=source, path=path, language_hint=language_hint, budget=budget)

File: src/parsers/parserRegistry.ts
/**
 * Parser registry — faithful port of core/parsers/parser_registry.py.
 * Hand-written production module (protected).
 */
import * as py from "../runtime/pyCompat.js";
import { parseAst } from "./astEngine.js";
import { buildCallGraph } from "./callGraphEngine.js";
import { resolveDependencies } from "./dependencyResolutionEngine.js";
import { resolveImports } from "./importResolutionEngine.js";
import { ParserBudget, enforceBudget } from "./parserBudgetEngine.js";
import { recoverSyntax } from "./parserRecoveryEngine.js";
import { resolveRuntime } from "./runtimeResolutionEngine.js";
import { resolveApiSurface } from "./apiResolutionEngine.js";
import { resolveFrameworks } from "./frameworkResolutionEngine.js";
import { buildSemanticGraph } from "./semanticGraphEngine.js";
import { resolveSymbols } from "./symbolResolutionEngine.js";
import { normalizeParserOutput } from "./parserOutputEngine.js";
import { requireParserEvidence } from "./formalParserGroundingEngine.js";

const _EXTENSION_LANGUAGE: Record<string, string> = {
  ".py": "python",
  ".js": "javascript",
  ".jsx": "javascript",
  ".ts": "typescript",
  ".tsx": "typescript",
  ".java": "java",
  ".kt": "kotlin",
  ".go": "go",
  ".rs": "rust",
  ".dart": "dart",
};

export class ParserRegistry {
  static detect_language(path = "", hint = ""): string {
    if (py.truthy(hint)) return String(hint).toLowerCase();
    const ext = py.path(path).suffix.toLowerCase();
    return _EXTENSION_LANGUAGE[ext] ?? "text";
  }

  static parse(
    source: string,
    path = "",
    language_hint = "",
    budget: ParserBudget | null = null,
  ): Record<string, unknown> {
    const language = ParserRegistry.detect_language(path, language_hint);
    const bounded = enforceBudget(source, budget);
    const recovered = recoverSyntax(bounded, language);
    const astTree = parseAst(recovered, language) as Record<string, unknown>;
    const symbols = resolveSymbols(recovered, language, astTree) as Record<string, unknown>;
    const calls = buildCallGraph(recovered, language) as Record<string, unknown>;
    const imports = resolveImports(symbols, py.truthy(path) ? path : language) as Record<string, unknown>;
    const deps = resolveDependencies(recovered, path) as Record<string, unknown>;
    const runtime = resolveRuntime(
      py.get(deps, "dependencies", []) as unknown[],
      py.get(symbols, "imports", []) as unknown[],
    ) as Record<string, unknown>;
    const frameworks = resolveFrameworks(
      py.get(deps, "dependencies", []) as unknown[],
      py.get(symbols, "imports", []) as unknown[],
      py.get(symbols, "decorators") as unknown,
    ) as Record<string, unknown>;
    const apiSurface = resolveApiSurface(recovered, language, path) as Record<string, unknown>;
    const parsed: Record<string, unknown> = {
      language,
      source_id: py.truthy(path) ? path : language,
      ast: astTree,
      symbols,
      calls,
      imports,
      dependencies: deps,
      runtime,
      frameworks,
      api_surface: apiSurface,
    };
    parsed["semantic_graph"] = buildSemanticGraph(parsed);
    const evidence: Record<string, unknown> = {
      ast: py.truthy(py.get(astTree, "nodes")),
      symbols: py.truthy(py.get(symbols, "symbols")),
      calls: py.truthy(py.get(calls, "calls")),
      dependencies: py.truthy(py.get(deps, "dependencies")),
      tree_sitter: language !== "python" && !py.truthy(py.get(astTree, "parse_error")),
    };
    parsed["evidence"] = evidence;
    if (py.truthy(py.get(astTree, "parse_error"))) {
      evidence["parse_error"] = true;
    }
    const normalized = normalizeParserOutput(parsed) as Record<string, unknown>;
    Object.assign(parsed, normalized);
    parsed["parser_grounding"] = requireParserEvidence(parsed);
    return parsed;
  }
}

export function parseSource(
  source: string,
  path = "",
  language_hint = "",
  budget: ParserBudget | null = null,
): Record<string, unknown> {
  return ParserRegistry.parse(source, path, language_hint, budget);
}

/* legacy registry surface (kept for compatibility) */
const REGISTRY = new Map<string, Record<string, unknown>>();

export function registerParser(name: string, meta: Record<string, unknown>): void {
  REGISTRY.set(name, { ...meta, name, bounded: true });
}

export function listParsers(): Array<Record<string, unknown>> {
  return [...REGISTRY.values()].sort((a, b) => (String(a.name) < String(b.name) ? -1 : 1));
}

export function getParser(name: string): Record<string, unknown> | undefined {
  return REGISTRY.get(name);
}
